File: app/utils/completeness.py
"""
app.utils.completeness — shared "empty" predicate and assessment completeness helper.

Used by:
  - LLM enrichment runner (logs warnings when assessment is sparse — no rejection)
  - tests for FORM_DATA_KEYS allowlist sanity
  - backend equivalent of frontend assessmentCompleteness for any future
    server-side gate (out of scope this delta but keeps the door open).

The "empty" predicate (design R4):
  None | "" | empty list/tuple/set/dict → empty.
  False, 0, 0.0 → FILLED (the user said "no" / "zero hours").

Whitespace-only strings are NOT trimmed before the check — by design.
The wizard already strips required fields via the not_empty validator in
PublicSubmissionPayload; non-required fields default to "".  A field returning
"  " is treated as filled (the user typed something).
"""
from __future__ import annotations

from typing import Any

from app.schemas.assessment import FORM_DATA_KEYS

# Critical form_data keys whose absence forces isSparse=True regardless of
# overall fill percentage (design §6, CAP-P-IN-001).
CRITICAL_FORM_DATA_KEYS: frozenset[str] = frozenset({
    "most_time_consuming_process",
    "software_used",
    "daily_queries",
    "urgency",
})

# Critical flat-column fields — checked via the flat_fields dict argument.
CRITICAL_FLAT_FIELDS: frozenset[str] = frozenset({"sector"})

# Fill-rate threshold below which the assessment is considered sparse.
# 60% of 42 = at least 26 of 42 filled required to pass.
COMPLETENESS_THRESHOLD: float = 0.6


def is_empty(value: Any) -> bool:
    """R4: None | "" | [] | {} are empty; False, 0, 0.0 are filled.

    Examples:
        is_empty(None)     → True
        is_empty("")       → True
        is_empty([])       → True
        is_empty({})       → True
        is_empty(False)    → False  # explicit "no" answer
        is_empty(0)        → False  # explicit "zero hours" answer
        is_empty(0.0)      → False
        is_empty("hello")  → False
        is_empty([1])      → False
    """
    if value is None:
        return True
    if isinstance(value, str) and value == "":
        return True
    if isinstance(value, (list, tuple, set, dict)) and len(value) == 0:
        return True
    return False


def assessment_completeness(
    form_data: dict[str, Any] | None,
    flat_fields: dict[str, Any],
) -> dict[str, Any]:
    """Compute completeness for an assessment.

    Args:
        form_data:    The assessment.form_data blob (or None).
        flat_fields:  Subset of flat columns relevant to the gate.
                      Must include 'sector' to satisfy CRITICAL_FLAT_FIELDS.

    Returns:
        dict with keys:
          filled          — number of canonical form_data keys that are filled
          total           — len(FORM_DATA_KEYS) == 42
          percentage      — filled / total  (0.0 .. 1.0)
          missing_critical — list of critical field names that are empty
          is_sparse       — True when percentage < threshold OR any critical missing
    """
    fd = form_data or {}
    filled = sum(1 for k in FORM_DATA_KEYS if not is_empty(fd.get(k)))
    total = len(FORM_DATA_KEYS)
    percentage = filled / total if total else 0.0

    missing: list[str] = []
    for k in sorted(CRITICAL_FORM_DATA_KEYS):
        if is_empty(fd.get(k)):
            missing.append(k)
    for k in sorted(CRITICAL_FLAT_FIELDS):
        if is_empty(flat_fields.get(k)):
            missing.append(k)

    is_sparse = percentage < COMPLETENESS_THRESHOLD or bool(missing)
    return {
        "filled": filled,
        "total": total,
        "percentage": percentage,
        "missing_critical": missing,
        "is_sparse": is_sparse,
    }

File: frontend/src/admin/utils/completeness.ts
/**
 * completeness.ts — shared "empty" predicate and assessment completeness helper.
 *
 * Mirrors app/utils/completeness.py (design R4 + §6 + §7).
 * The predicate semantics MUST stay identical to the Python side.
 *
 * R4 contract:
 *   null | undefined | ""    → empty
 *   []                       → empty  (no items selected)
 *   {}                       → empty  (empty object — treated like missing)
 *   false                    → FILLED (explicit "no" answer)
 *   0                        → FILLED (explicit "zero hours" answer)
 *   0.0 (number)             → FILLED
 *
 * Whitespace-only strings are NOT trimmed — "  " is filled by design.
 */

import type { AssessmentResponse } from '../api/assessments'
import { FORM_DATA_KEYS } from './formDataKeys'

// ---------------------------------------------------------------------------
// Critical fields (must be filled to avoid isSparse regardless of fill rate)
// ---------------------------------------------------------------------------

/** Critical form_data keys — absence overrides fill-rate threshold. */
const CRITICAL_FORM_DATA: readonly string[] = [
  'most_time_consuming_process',
  'software_used',
  'daily_queries',
  'urgency',
] as const

/** Critical flat-column keys on AssessmentResponse. */
const CRITICAL_FLAT: readonly (keyof AssessmentResponse)[] = [
  'sector',
] as const

/** Fill-rate threshold: < 60% (26/42) → sparse. */
const COMPLETENESS_THRESHOLD = 0.6

// ---------------------------------------------------------------------------
// isEmpty predicate
// ---------------------------------------------------------------------------

/**
 * Returns true when `value` is considered "empty" per design R4.
 *
 * @example
 *   isEmpty(null)      // true
 *   isEmpty(undefined) // true
 *   isEmpty('')        // true
 *   isEmpty([])        // true
 *   isEmpty({})        // true
 *   isEmpty(false)     // false  ← explicit "no" answer
 *   isEmpty(0)         // false  ← explicit "zero" answer
 *   isEmpty('hello')   // false
 *   isEmpty([1])       // false
 */
export function isEmpty(value: unknown): boolean {
  if (value == null) return true
  if (typeof value === 'string' && value === '') return true
  if (Array.isArray(value) && value.length === 0) return true
  if (
    typeof value === 'object' &&
    value !== null &&
    !Array.isArray(value) &&
    Object.keys(value as Record<string, unknown>).length === 0
  ) {
    return true
  }
  return false
}

// ---------------------------------------------------------------------------
// assessmentCompleteness
// ---------------------------------------------------------------------------

export interface CompletenessResult {
  /** Number of canonical form_data keys that have a filled value. */
  filledKeys: number
  /** Total canonical keys — always FORM_DATA_KEYS.length (42). */
  totalKeys: number
  /** filledKeys / totalKeys (0..1). */
  percentage: number
  /** Critical field names that are empty (form_data keys + flat keys). */
  missingCritical: string[]
  /** True when percentage < 0.6 OR any critical field is missing. */
  isSparse: boolean
}

/**
 * Compute completeness for an assessment.
 *
 * Mirrors Python `assessment_completeness(form_data, flat_fields)`.
 *
 * @param assessment - Full AssessmentResponse (or undefined for safe zero-state).
 * @returns CompletenessResult
 */
export function assessmentCompleteness(
  assessment: AssessmentResponse | undefined,
): CompletenessResult {
  if (assessment == null) {
    return {
      filledKeys: 0,
      totalKeys: FORM_DATA_KEYS.length,
      percentage: 0,
      missingCritical: [...CRITICAL_FORM_DATA, ...CRITICAL_FLAT],
      isSparse: true,
    }
  }

  const fd = (assessment.form_data ?? {}) as Record<string, unknown>

  const filledKeys = FORM_DATA_KEYS.filter((k) => !isEmpty(fd[k])).length
  const totalKeys = FORM_DATA_KEYS.length
  const percentage = filledKeys / totalKeys

  const missingCritical: string[] = []
  for (const k of CRITICAL_FORM_DATA) {
    if (isEmpty(fd[k])) missingCritical.push(k)
  }
  for (const k of CRITICAL_FLAT) {
    if (isEmpty(assessment[k])) missingCritical.push(k as string)
  }

  const isSparse = percentage < COMPLETENESS_THRESHOLD || missingCritical.length > 0

  return { filledKeys, totalKeys, percentage, missingCritical, isSparse }
}
